### PR TITLE
[AArch64][CostModel] Improve cost estimate of scalarizing a vector di…

### PR DESCRIPTION
**Pull request by sushgokh**
…vision

In the backend, last resort of finding the vector division cost is to use its scalar cost. However, without knowledge about the division operands, the cost can be off in certain cases.

For SLP, this patch tries to pass scalars for better scalar cost estimation in the backend.

Tested the patch on Neoverse-v2 with SPEC17(C/C++). No regressions observed.

### CODE REVIEW

File: llvm/test/Analysis/CostModel/AArch64/div.ll [hunk 1]
```diff
@@ -7,21 +7,21 @@ define i32 @sdiv() {
 ; CHECK-LABEL: 'sdiv'
 ; CHECK-NEXT:  Cost Model: Found an estimated cost of 10 for instruction: %I128 = sdiv i128 undef, undef
 ; CHECK-NEXT:  Cost Model: Found an estimated cost of 1 for instruction: %I64 = sdiv i64 undef, undef
-; CHECK-NEXT:  Cost Model: Found an estimated cost of 28 for instruction: %V2i64 = sdiv <2 x i64> undef, undef
```
Comment (david-arm) [inline llvm/test/Analysis/CostModel/AArch64/div.ll:10]: It feels like this patch is doing two things at once:

1. Changing the costs of divides when not passing in the new Scalars parameter,
2. Changing the costs of divides when passing in a non-null Scalars.

It's probably worth splitting the changes up into two different PRs so we can see review them independently. It also allows us to monitor the performance impact of each change on it's own.

File: llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp [hunk 2]
```diff
@@ -3472,29 +3472,41 @@ InstructionCost AArch64TTIImpl::getArithmeticInstrCost(
...
+              (Op2Info.isConstant() && Op2Info.isUniform())) {
             InstructionCost DivCost = BaseT::getArithmeticInstrCost(
                 Opcode, Ty->getScalarType(), CostKind, Op1Info, Op2Info);
-            return (4 + DivCost) * VTy->getNumElements();
+            // If #vector_elements = n then we need
+            // n inserts + 2n extracts + n divisions.
+            InstructionCost InsertExtractCost =
+                ST->getVectorInsertExtractBaseCost();
+            Cost = (3 * InsertExtractCost + DivCost) * VTy->getNumElements();
+          } else if (!Scalars.empty()) {
```
Comment (alexey-bataev) [inline llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:3485]: Why Args is not enough for this analysis?

File: llvm/test/Analysis/CostModel/AArch64/div.ll (hunk 1 above)
Comment (sushgokh) [reply llvm/test/Analysis/CostModel/AArch64/div.ll:10]: changes in `llvm/test/Analysis/CostModel/AArch64/div.ll` are not completely orthogonal to my patch (or (1) is not completely orthogonal to (2) ).


[This](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp#L3479C1-L3486C10) piece of code needs change for following reasons:
1. Now that I have extra information about the scalars and information about the scalar type can be passed  i.e. instead of passing Op1Info and Op2Info, I need to pass information about the scalar. I forgot to update this. Will update the patch.
2. The extract cost, which contributes to the entire cost, is considered once instead of twice [here](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp#L3484) which is not correct I think. 

In short, updating this small piece of code would cause changes in `llvm/test/Analysis/CostModel/AArch64/div.ll`

The perf results, as mentioned in the description, are neutral.

File: llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp (hunk 2 above)
Comment (alexey-bataev) [reply llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:3485]: Need to investigate this rather than adding some extra args

Comment (sushgokh) [reply llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp]: done

File: llvm/test/Analysis/CostModel/AArch64/div.ll (hunk 1 above)
Comment (sushgokh) [reply llvm/test/Analysis/CostModel/AArch64/div.ll]: @david-arm segregated two things as you mentioned.

**Comment by sushgokh** [comment]
ping

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 3]
```diff
@@ -11561,9 +11561,9 @@ BoUpSLP::getEntryCost(const TreeEntry *E, ArrayRef<Value *> VectorizedVals,
       unsigned OpIdx = isa<UnaryOperator>(VL0) ? 0 : 1;
       TTI::OperandValueInfo Op1Info = getOperandInfo(E->getOperand(0));
       TTI::OperandValueInfo Op2Info = getOperandInfo(E->getOperand(OpIdx));
-      return TTI->getArithmeticInstrCost(ShuffleOrOp, VecTy, CostKind, Op1Info,
-                                         Op2Info, {}, nullptr, TLI) +
-             CommonCost;
+      return CommonCost +
+             TTI->getArithmeticInstrCost(ShuffleOrOp, VecTy, CostKind, Op1Info,
+                                         Op2Info, E->Scalars, nullptr, TLI);
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:11566]: This is incorrect, it should pass operands, not the original scalars

Comment (sushgokh) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:11566]: But, how does it matter if we pass scalars instead of operands here?

Comment (alexey-bataev) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:11566]: It breaks the the API contract

Comment (sushgokh) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp]: resolved

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 4]
```diff
@@ -11561,9 +11561,25 @@ BoUpSLP::getEntryCost(const TreeEntry *E, ArrayRef<Value *> VectorizedVals,
...
-      return TTI->getArithmeticInstrCost(ShuffleOrOp, VecTy, CostKind, Op1Info,
-                                         Op2Info, {}, nullptr, TLI) +
-             CommonCost;
+      SmallVector<Value *, 16> Operands;
+      auto IsAllowedScalarTy = [](const Value *I) {
+        // Just to avoid regression with RISCV unittest.
+        return I->getType()->getPrimitiveSizeInBits() >= 32;
+      };
+      if (all_of(E->Scalars, [ShuffleOrOp, IsAllowedScalarTy](Value *V) {
+            return !IsaPred<UndefValue, PoisonValue>(V) &&
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:11570]: `return !isa<UndefValue, PoisonValue>(V) &&`

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 5]
```diff
@@ -11561,9 +11561,25 @@ BoUpSLP::getEntryCost(const TreeEntry *E, ArrayRef<Value *> VectorizedVals,
...
       TTI::OperandValueInfo Op1Info = getOperandInfo(E->getOperand(0));
       TTI::OperandValueInfo Op2Info = getOperandInfo(E->getOperand(OpIdx));
-      return TTI->getArithmeticInstrCost(ShuffleOrOp, VecTy, CostKind, Op1Info,
-                                         Op2Info, {}, nullptr, TLI) +
-             CommonCost;
+      SmallVector<Value *, 16> Operands;
+      auto IsAllowedScalarTy = [](const Value *I) {
+        // Just to avoid regression with RISCV unittest.
+        return I->getType()->getPrimitiveSizeInBits() >= 32;
+      };
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:11568]: Hacks are the bad way.

Comment (sushgokh) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:11568]: These Args are being used by the  RISCV target for calculating the cast cost. To avoid miscalculation with the patch, had to use this hack.
If we dont want this, we need to introduce other parameter. There's no way out here.

Comment (alexey-bataev) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:11568]: Or adjust RISCVTargetTransformInfo.cpp to handle it correctly, if there is something wrong, or update RISCV tests, if everything is fine

Comment (sushgokh) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:11568]: adjusting RISCVTargetTransformInfo.cpp or updating the test & justifying the results is beyond my scope . w/o patch, casting is done only for scalar. w/ patch, its being done for all operands of the all the scalars.

Comment (alexey-bataev) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:11568]: > adjusting RISCVTargetTransformInfo.cpp or updating the test & justifying the results is beyond my scope . w/o patch, casting is done only for scalar. w/ patch, its being done for all operands of the all the scalars.

In upstream you should care about all targets, you cannot introduce hacks just for your target. You can do this in your downstream version.

Comment (sushgokh) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:11568]: Thats the reason I wanted to introduce one more API parameter

Comment (sushgokh) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp]: I have tried to debug the RISCV backend and looking at the comment
`// f16 with zvfhmin and bf16 will be promoted to f32.`, calculating the cast cost for the args seems a valid choice which isnt happening currently.

But if this is incorrect, could you please help me to avoid this regression?

**Comment by alexey-bataev** [comment]
Do not use force push

**Review by alexey-bataev — CHANGES_REQUESTED** [review]

File: llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp [hunk 6]
```diff
@@ -3472,6 +3472,20 @@ InstructionCost AArch64TTIImpl::getArithmeticInstrCost(
...
+          InstructionCost InsertExtractCost =
+              ST->getVectorInsertExtractBaseCost();
+          Cost = (3 * InsertExtractCost) * VTy->getNumElements();
+          for (int i = 0, Sz = Args.size(); i < Sz; i += 2) {
+            Cost += getArithmeticInstrCost(
+                Opcode, VTy->getScalarType(), CostKind,
+                TTI::getOperandInfo(Args[i]), TTI::getOperandInfo(Args[i + 1]));
+          }
+          return Cost;
+        }
```
[the remark this answers is not in the document]
Comment (sushgokh) [reply llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:3487]: > I don't understand what are you trying to do here. Why are you adding the cost of the scalar ops here? They are already counted in the cost model

AArch64 Neon does not support vector division natively. It scalarizes the vector division as can be seen [here](https://godbolt.org/z/EaxKsGe3b). 

Now coming to the cost modelling part: This code section is calculating the cost of scalarizing the vector division. Thus,
```
vector cost = 
for each lane(cost of extract for both div operands) + 
for each lane(cost of scalar division) + 
for each lane(cost of insert into the result vector) 
```
This vector cost was being over-calculated thus preferring scalar code. This patch just tries to rectify this.

[the remark this answers is not in the document]
Comment (alexey-bataev) [reply llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:3487]: Then, you should not build the vector node at all. I would suggest instead that you check how the `getScalarsVectorizationState` function in SLP works, add a check for NEON division (maybe add a new entry in TTI to check if the vector operation is legal and won't be scalarized), and return that for NEON, it should build the `TreeEntry::NeedToGather` node.

[the remark this answers is not in the document]
Comment (sushgokh) [reply llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:3487]: > Then, you should not build the vector node at all. I would suggest instead that you check how the `getScalarsVectorizationState` function in SLP works, add a check for NEON division (maybe add a new entry in TTI to check if the vector operation is legal and won't be scalarized), and return that for NEON, it should build the `TreeEntry::NeedToGather` node.

Two issues which I can think of with this approach:
1. The operands of the div operation wont get vectorized.
2. Lot of checks are in place that alter the div cost based on information about  its operands(e.g. if operands are uniform/constant/pow-of-2 etc.). Hence, factoring out code just to check if the div operation would be scalarized is difficult and would result in duplication of code.

[the remark this answers is not in the document]
Comment (sushgokh) [reply llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:3487]: > 1. That's fine. If you think they should be vectorized, a new vectorization strategy should be implemented instead.

Don't you think lowering the division vector in some late IR pass is easy than stopping vectorization at div ? One more advantage of lowering later is lowering is transparent to the user and done at single place rather than at multiple places.
This type of vector lowering is already being done at other places e.g. masked_load

> 2\. No, it is just a simple legality check. If the instruction should be scalarized, it should be built into a build vector node; it should not be vectorized.

Just need a clarification: my understanding of `TreeEntry::NeedToGather` is you build a vector out of scalars. Is that correct?
Now, coming to the legality checks. Yes, in the end, its legality check but I need to check all the  contraints, which are currently in `getArithmeticInstrCost` once again. And I am trying to avoid that route just for reasons mentioned in my comment on (1).

[the remark this answers is not in the document]
Comment (alexey-bataev) [reply llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:3487]: > > 1. That's fine. If you think they should be vectorized, a new vectorization strategy should be implemented instead.
> 
> Don't you think lowering the division vector in some late IR pass is easy than stopping vectorization at div ? One more advantage of lowering later is lowering is transparent to the user and done at single place rather than at multiple places. This type of vector lowering is already being done at other places e.g. masked_load
> 

It just breaks the vectorizer design. If the node should be scalarized, the vectorization model should know about it and scalarize it. Otherwise, it complicates the whole vectorization process and makes it miss some possible vectorization opportunities. I assume, you missed the cost of extraction of the operands in your cost model changes?

> > 2. No, it is just a simple legality check. If the instruction should be scalarized, it should be built into a build vector node; it should not be vectorized.
> 
> Just need a clarification: my understanding of `TreeEntry::NeedToGather` is you build a vector out of scalars. Is that correct? Now, coming to the legality checks. Yes, in the end, its legality check but I need to check all the contraints, which are currently in `getArithmeticInstrCost` once again. And I am trying to avoid that route just for reasons mentioned in my comment on (1).

You can check these constraints before and command vectorizer to build a NeedToGather node. You should not use getArithmeticInstrCost in this case, some other function should check if it is legal to vectorize these scalars here

[the remark this answers is not in the document]
Comment (sushgokh) [reply llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:3487]: > Otherwise, it complicates the whole vectorization process and makes it miss some possible vectorization opportunities. I assume, you missed the cost of extraction of the operands in your cost model changes?

No, in fact the patch is accurately calculating the cost of:
```
for every lane: 2 extracts and 1 div and 1 insert
See
   Cost = (3 * InsertExtractCost) * VTy->getNumElements();   // Insert/Extract cost is same
```
So, no way it is missing out on opportunities and by no means its complicating the process.

**Comment by sushgokh** [comment]
ping

**Comment by davemgreen** [comment]
I happened to be looking at urem costs recently - not exactly related to this but led to a similar area. I was only going to update the urem by constant values but the sdiv costs are similar.

I'm not sure this needs to alter the SLP vectorizer for the example tests. So long as all the second operands are constant or power-of-two it should be expanding to a not use a div (and not scalarize). I think we can update this bit of code:
https://github.com/llvm/llvm-project/blob/2b26ee6e790574e05c3c9a562bc37897daf0f384/llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp#L3495
 - It probably should not matter that the constant is uniform.
 - It should handle negative-powers of 2 too for sdiv.
 - The cost should use Add + ICmp + Sel + AShr. (The sel goes through the wrong method).
 - We can maybe add SVE asrd but that can happen in a later patch.
 - I believe vector types have the same expansion and costs as scalars.
 - srem has a similar cost to sdiv for these values, but that should probably be added later too, as there is more than just power-2 values.

**Comment by davemgreen** [comment]
The urem patch I mentioned is now in #122236.

**Comment by sushgokh** [comment]
> The urem patch I mentioned is now in #122236.

Thanks for the patch. This serves as direction for sdiv/srem and will take this up next.

Although it resolves the motivating case, it does not resolve the generic case where Op2 is unknown. Consider below example(motivating  example modified with sdiv->udiv and divisor changed to unknowns)
```
define <2 x i32> @vectorize_udiv_v2i32(<2 x i32> %a, <2 x i32> %x, <2 x i32> %y, <2 x i32> %z)
{
  %a0 = extractelement <2 x i32> %a, i64 0
  %a1 = extractelement <2 x i32> %a, i64 1
  %1 = udiv i32 %a0, %a1
  %2 = udiv i32 %a1, %a0
  %x0 = extractelement <2 x i32> %x, i64 0
  %x1 = extractelement <2 x i32> %x, i64 1
  %3 = add i32 %1, %x0
  %4 = add i32 %2, %x1
  %y0 = extractelement <2 x i32> %y, i64 0
  %y1 = extractelement <2 x i32> %y, i64 1
  %5 = sub i32 %3, %y0
  %6 = sub i32 %4, %y1
  %z0 = extractelement <2 x i32> %z, i64 0
  %z1 = extractelement <2 x i32> %z, i64 1
  %7 = mul i32 %5, %z0
  %8 = mul i32 %6, %z1
  %res0 = insertelement <2 x i32> poison, i32 %7, i32 0
  %res1 = insertelement <2 x i32> %res0, i32 %8, i32 1
  ret <2 x i32> %res1
}
```

urem patch fails to vectorize this while this patch vectorizes this.

In summary, this patch is necessary as it refines the scalar cost. 

Last thing: I will update the motivating case as above.

File: llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp [hunk 7]
```diff
@@ -18,6 +18,7 @@
 #include "llvm/CodeGen/BasicTTIImpl.h"
 #include "llvm/CodeGen/CostTable.h"
 #include "llvm/CodeGen/TargetLowering.h"
+#include "llvm/IR/Constants.h"
```
Comment (sushgokh) [inline llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:21]: will remove this in next version. Got this accidently

**Comment by sushgokh** [comment]
ping

File: llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp [hunk 8]
```diff
@@ -3572,6 +3573,40 @@ InstructionCost AArch64TTIImpl::getArithmeticInstrCost(
...
+        if (auto *VTy = dyn_cast<FixedVectorType>(Ty); VTy && !Args.empty()) {
+          assert(Args.size() % 2 == 0 && "Args size should be even");
+          InstructionCost InsertExtractCost =
+              ST->getVectorInsertExtractBaseCost();
+          // If the cost of single sdiv is inquired through the cost-model.
+          // FIXME: remove the isa checks once the PR 122236 lands.
+          if (Args.size() == 2 &&
+              !(isa<ConstantVector>(Args[1]) ||
+                isa<ConstantDataVector>(Args[1]) ||
+                isa<ConstantExpr>(Args[1])) &&
```
Comment (alexey-bataev) [inline llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:3587]: I assume you do not need to pass args for this, instead you can rely on Op1Info and Op2Info

Comment (sushgokh) [reply llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:3587]: for the test where Op1 only or Op2 only is constant(see the added tests), simply relying on Op1Info and Op2Info does not work. In such cases, we need to pass args

Comment (alexey-bataev) [reply llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:3587]: Why?

Comment (sushgokh) [reply llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:3587]: the cost of the division when the divisor is constant is less than the case where the divisor is unknown.

When we are considering scalarizing cost, we are considering div cost of each lane and additional insert/extract cost. This is where this patch yeilds less cost with extra knowledge about the values that go into that lanes

Comment (alexey-bataev) [reply llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:3587]: Also, I told this before already, if the node is scalarized, it must be represented as a buildvector node.

Comment (sushgokh) [reply llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:3587]: Please check the case `@sdiv_v2i32_Op1_unknown_Op2_const` where only one of the divisors is constant. In that  case, Op2Info would be something generic and doesnt give sufficient information. I hope this makes sense

Comment (sushgokh) [reply llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:3587]: > Also, I told this before already, if the node is scalarized, it must be represented as a buildvector node.

I have tried to justify this why this wont be a better option already

**Review by alexey-bataev — CHANGES_REQUESTED** [review]
Scalarized nodes must be represented as buildvectors. If the operands should be vectorized, need to add special processing for such operands. Vectorizing `must be scalarized` nodes comes against design of the vectorizer.
